### PR TITLE
Add API Annotations to Ludwig

### PR DESCRIPTION
**Pull request by arnavgarg1**
This PR introduces `api_annotations.py` that can be used to delineate the boundary of Ludwig APIs that are public, for developers, and those that are deprecated.

Co-authored-by: Daniel Treiman <daniel@predibase.com> (@dantreiman)

### CODE REVIEW

File: ludwig/api_annotations.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+    def inner(obj):
+        _append_doc(obj, message=message, directive="warning")
+        _mark_annotated(obj)
+        return obj
+
+    return inner
+
+
+def _append_doc(obj, *, message: str, directive: Optional[str] = None) -> str:
```
Comment (dantreiman) [inline ludwig/api_annotations.py:119]: Also I'm not sure why there is a * here in the arg list, I don't see anywhere we pass in more arguments.  how about
```python
def _append_doc(obj, message: str, directive: Optional[str] = None) -> str:
```

**Review by dantreiman — CHANGES_REQUESTED** [review]

File: ludwig/api_annotations.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+    return inner
+
+
+def _append_doc(obj, message: str, directive: Optional[str] = None) -> str:
+    """
+    Args:
+        message: An additional message to append to the end of docstring for a class
+                 or method that uses one of the API annotations
+        directive:
```
Comment (arnavgarg1) [inline ludwig/api_annotations.py:115]: @dantreiman Do you know what the `directive` argument does exactly? I can see that you can pass in things like `warning` or `exception`, but just want to make sure I understand this right.

Comment (dantreiman) [reply ludwig/api_annotations.py:112]: AFAICT it just prepends it on a line of the docstring and indents the message.


**directive is not None**
```
directive:  
    message
```

**directive is None**
```
message
```

**Review by dantreiman — APPROVED** [review]

File: ludwig/api_annotations.py (hunk 2 above)
Comment (arnavgarg1) [reply ludwig/api_annotations.py:112]: Yup that's what I could tell as well, sounds good!

File: ludwig/api_annotations.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    Args:
+        stability: One of {"stable", "experimental"}
+
+    Examples:
+        >>> from api_annotations import PublicAPI
+        >>> @PublicAPI
+        ... def func1(x):
+        ...     return x
+        >>> @PublicAPI(stability="beta")
```
Comment (jppgks) [inline ludwig/api_annotations.py:21]: we define the choices above as `{"stable", "experimental"}`, here we mention `"beta"` too

**Review by jppgks — APPROVED** [review]

File: ludwig/api_annotations.py (hunk 3 above)
Comment (arnavgarg1) [reply ludwig/api_annotations.py]: Great catch, just fixed!

**Review by justinxzhao — APPROVED** [review]

**Pull request merged**